### ISSUE TITLE
Per face materials for CSG

### ISSUE THREAD

**Issue by Zireael07** [godotengine/godot-proposals#937]
**Describe the project you are working on:**
3d racing game that makes heavy use of CSG because of it's lowpoly nature

**Describe the problem or limitation you are having in your project:**
CSG materials are all or nothing. Sometimes they fit perfectly on one face, but are stretched on another.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Per face CSG materials would massively increase the utility of CSG shapes.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Mockup in original issue.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Nope.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Requires changes in CSG module.

Original issue: https://github.com/godotengine/godot/issues/22559

**Comment by ElfEars** [comment]
Thanks for posting this @Zireael07 .
I totally support this as (as previously mentioned) the current material system is extremely limited and others who've attempted to remedy this have had to completely re-implement CSG.

Here's the diagram from the old post (Which I stole and modified from a quake mapping tutorial)
![46259095-762f8b80-c4cc-11e8-9a39-6bd464fa1cd3](https://user-images.githubusercontent.com/20967774/84598060-e4161500-ae5f-11ea-9e6c-259f69ac8670.jpg)


Here's a (crude) mock-up of how an editor GUI for such a system could work. 
Feedback would be appreciated but I think it gets most things down.
![CSGProposalv3](https://user-images.githubusercontent.com/20967774/84598014-9ef1e300-ae5f-11ea-9abf-044085068b81.png)

**Comment by ElfEars** [comment]
It's also worth mentioning that texture scale should be in the GLOBAL context
IE: If you scale a mesh, the textures should continue tiling instead of scaling with the mesh

**Comment by drequivalent** [comment]
> It's also worth mentioning that texture scale should be in the GLOBAL context
> IE: If you scale a mesh, the textures should continue tiling instead of scaling with the mesh

Or have a texture lock switch. Like Trenchbroom has.

**Comment by NewNodeGames** [comment]
I tried Trenchbroom and FreeCad and no one is so intuitive and easy to use as Godot csg system and is integrated in the engine. I really like it but is a bit limited and sometimes the result can broke and has texture or shading problems with smooth faces. Would be very nice features like this.

**Comment by Calinou** [comment]
I looked into the CSG code recently and it seems that it's already possible to have a single CSG node having multiple materials. For the CSG primitives, we could expose the following:

- CSGBox: 1 material per face (6 total)
- CSGCylinder: 1 material for the top, 1 material for the sides, 1 material for the bottom (3 total)
- CSGSphere: 1 material (like currently)
- CSGPolygon: 1 material for the top, 1 material for the sides, 1 material for the bottom (3 total)
- CSGTorus: 1 material (like currently)
- CSGMesh: 1 material, unless we figure out a way to extract the mesh's material slots which would allow for as many materials as the mesh exposes

The main problem with doing this is that it would require much more clicks to assign a material to all a CSG node's faces.

**Comment by Zireael07** [comment]
Regarding the number of clicks, would it be possible to default to one and the same material for all faces and/or have a button somewhere that says 'apply to all faces'?

**Comment by Calinou** [comment]
> Regarding the number of clicks, would it be possible to default to one and the same material for all faces and/or have a button somewhere that says 'apply to all faces'?

I guess we could make faces with unset materials use the first material, which effectively means other materials would be face-specific overrides.

**Comment by novemberist** [comment]
With CSGs finally getting some love thanks to the great work of @fire, I hope this proposal also gets re-evaluated at some point. Lacking flexibility when texturing is the one thing that makes working with CSGs still kind of cumbersome. Seems like someone was already halfway there at some point in https://github.com/godotengine/godot/pull/59266, but unfortunately this topic seems to have lost traction.

**Comment by fire** [comment]
I’m probably more interested in getting skeletons and weighting working, but I can review any prs made.

**Comment by novemberist** [comment]
Fair enough. I suppose decals are a somewhat viable alternative to put textures on individual faces of simple csg meshes for prototyping. If only there was a way to somehow make decal textures tileable and cull masks work on csg child nodes.

**Comment by Radivarig** [comment]
This is possible with a thin slice of subtract operation that has a different material set

<img width="923" height="402" alt="Image" src="https://github.com/user-attachments/assets/13cef7c2-12c5-4c63-b395-b6200ec83564" />
